### PR TITLE
Add epsilon predictor support to diffusion framework

### PR DESCRIPTION
**Pull request by CharlelieLrt**
# PhysicsNeMo Pull Request

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: physicsnemo/diffusion/metrics/losses.py
Comment (laserkelvin) [inline physicsnemo/diffusion/metrics/losses.py]: Personal style, but I think if there are three cases it's probably better to do

```python
match prediction_type:
    case "x0": ...
    case "score": ...
    case "epsilon": ...
    case _: raise Exception
```

File: physicsnemo/diffusion/metrics/losses.py [hunk 1]
```diff
@@ -327,11 +258,15 @@ def __init__(
         self,
         model: DiffusionModel,
         noise_scheduler: NoiseScheduler,
-        prediction_type: Literal["x0", "score"] = "x0",
+        prediction_type: Literal["x0", "score", "epsilon"] = "x0",
```
Comment (laserkelvin) [inline physicsnemo/diffusion/metrics/losses.py]: Maybe consider aliasing `prediction_type`, so that you can re-use it in the Loss `__init__`?

File: physicsnemo/diffusion/noise_schedulers/noise_schedulers.py [hunk 2]
```diff
@@ -769,19 +765,196 @@ def score_to_x0(
...
+        Examples
+        --------
+        >>> scheduler = EDMNoiseScheduler()
+        >>> eps = torch.randn(2, 4)
+        >>> t = torch.tensor([1.0, 1.0])
+        >>> score = scheduler.epsilon_to_score(eps, t)
+        >>> score.shape
+        torch.Size([2, 4])
+        """
+        t_bc = t.reshape(-1, *([1] * (epsilon.ndim - 1)))
```
Comment (laserkelvin) [inline physicsnemo/diffusion/noise_schedulers/noise_schedulers.py]: This is pretty hard to read IMO

File: physicsnemo/diffusion/noise_schedulers/noise_schedulers.py [hunk 3]
```diff
@@ -769,19 +765,196 @@ def score_to_x0(
...
+        Examples
+        --------
+        >>> scheduler = EDMNoiseScheduler()
+        >>> score = torch.randn(2, 4)
+        >>> t = torch.tensor([1.0, 1.0])
+        >>> eps = scheduler.score_to_epsilon(score, t)
+        >>> eps.shape
+        torch.Size([2, 4])
+        """
+        t_bc = t.reshape(-1, *([1] * (score.ndim - 1)))
```
Comment (laserkelvin) [inline physicsnemo/diffusion/noise_schedulers/noise_schedulers.py]: Same readability comment

File: physicsnemo/diffusion/noise_schedulers/noise_schedulers.py [hunk 4]
```diff
@@ -769,19 +765,196 @@ def score_to_x0(
...
+        """
+        t_bc = t.reshape(-1, *([1] * (x0.ndim - 1)))
+        alpha_t_bc = self.alpha(t_bc)
+        sigma_t_bc = self.sigma(t_bc)
+        return (x_t - alpha_t_bc * x0) / sigma_t_bc
+
     def get_denoiser(
         self,
         *,
         score_predictor: Predictor | None = None,
```
Comment (laserkelvin) [inline physicsnemo/diffusion/noise_schedulers/noise_schedulers.py:959]: Design-wise, I'm not sure if the interface would be better if it was:

```python
self,
predictor: Predictor,
predictor_type: Literal[...],
**kwargs: Any,
)
```

And then just implement the logic based on `predictor_type` with the same `match/case` pattern. Main reason is that they are mutually exclusive anyway, and I think this way makes it easier to support new types more cleanly.

**Review by laserkelvin — COMMENTED** [review]
Left some design considerations - let me know what you think.

File: physicsnemo/diffusion/noise_schedulers/noise_schedulers.py (hunk 4 above)
Comment (CharlelieLrt) [reply physicsnemo/diffusion/noise_schedulers/noise_schedulers.py:959]: Yeah right. Current interface is based on a suggestion from Nick. In normal circumstances, I would have no problem applying your suggestion, but that is breaking backward compatibility. I would prefer to prioritize API stability at the moment

Comment (laserkelvin) [reply physicsnemo/diffusion/noise_schedulers/noise_schedulers.py:959]: Makes sense - let's mark this as resolved

File: physicsnemo/diffusion/noise_schedulers/noise_schedulers.py (hunk 2 above)
Comment (CharlelieLrt) [reply physicsnemo/diffusion/noise_schedulers/noise_schedulers.py]: What would you suggest to improve readability?

**Review by laserkelvin — APPROVED** [review]
Approving

File: physicsnemo/diffusion/noise_schedulers/noise_schedulers.py (hunk 2 above)
Comment (laserkelvin) [reply physicsnemo/diffusion/noise_schedulers/noise_schedulers.py]:
```python
expected_shape = (-1,) + (1,) * (epsilon.ndim - 1)
t_bc = t.reshape(expected_shape)
```

Comment (laserkelvin) [reply physicsnemo/diffusion/noise_schedulers/noise_schedulers.py]: Understandably subjective :)

File: physicsnemo/diffusion/metrics/losses.py
Comment (CharlelieLrt) [reply physicsnemo/diffusion/metrics/losses.py]: Agree — replaced `if/elif` with `match/case` in `MSEDSMLoss` and `WeightedMSEDSMLoss`, and also applied it to `MultiDiffusionMSEDSMLoss` and `MultiDiffusionWeightedMSEDSMLoss` for consistency across the diffusion framework. Fixed in 4927061e.

File: physicsnemo/diffusion/metrics/losses.py (hunk 1 above)
Comment (CharlelieLrt) [reply physicsnemo/diffusion/metrics/losses.py]: Good suggestion — added `PredictorType = Literal["x0", "score", "epsilon"]` in [`physicsnemo/diffusion/base.py`](https://github.com/NVIDIA/physicsnemo/blob/4927061e/physicsnemo/diffusion/base.py#L25-L32) and used it in the `MSEDSMLoss` / `WeightedMSEDSMLoss` signatures and docstrings. The `multi_diffusion` losses keep the narrower `Literal["x0", "score"]` since they do not support epsilon yet. Fixed in 4927061e.

File: physicsnemo/diffusion/noise_schedulers/noise_schedulers.py (hunk 2 above)
Comment (CharlelieLrt) [reply physicsnemo/diffusion/noise_schedulers/noise_schedulers.py]: Agree — applied your suggested two-line form not only here but across every occurrence of this pattern in `physicsnemo/diffusion/` (noise_schedulers, guidance, samplers, preconditioners, plus doctest examples in `dps_guidance.py`). Lambdas in `samplers/solvers.py` are left as-is since the form does not fit there. Fixed in 4927061e.

File: physicsnemo/diffusion/noise_schedulers/noise_schedulers.py (hunk 3 above)
Comment (CharlelieLrt) [reply physicsnemo/diffusion/noise_schedulers/noise_schedulers.py]: Fixed in 4927061e along with all other occurrences of the same pattern — see reply on the line 810 thread for details.

**Review by ktangsali — APPROVED** [review]
